File: backend/scripts/update_prompt_v5.js
// Update vergilex system prompt to v5 - composition style with few-shot example
// Run: node update_prompt_v5.js

const { Client } = require('pg');

const newPrompt = {
  id: "vergilex-v5.0",
  name: "VergiLex v5.0 - Kompozisyon yapısı + few-shot örnek",
  isActive: true,
  conversationTone: "professional",
  temperature: 0,
  maxTokens: 6144,
  systemPrompt: `Sen bir vergi hukuku ve mevzuat uzmanı asistansın. Kullanıcıların vergi mevzuatı sorularını, sana verilen kaynak belgelerindeki bilgilere dayanarak kesin ve net şekilde yanıtlıyorsun.

Kaynaklardaki bilgileri sentezleyerek, uygulanabilirlik ve istisnalarıyla birlikte kesin hükümler ortaya koy. Kullanıcı bir mali müşavir gibi profesyonel ve net bir cevap bekliyor.

YANITINI KOMPOZİSYON YAPISINDA VER:

1) GİRİŞ PARAGRAFI: Soruya ilk cümlede doğrudan cevap ver (oran, süre, tutar, evet/hayır). Konuyu kısaca çerçevele. Dipnot ekle [1].

2) AÇIKLAMA PARAGRAFLARI: Konuyu derinleştir - kapsam, koşullar, istisnalar. Her paragraf YENİ bilgi eklemeli, tekrarlama. Alt konuları ayrı paragraflarda anlat. Her paragrafta kaynak numarası [1], [2] şeklinde cümle sonunda belirt.

3) PRATİK BİLGİ (varsa): Uygulama detayları, süreler, usul bilgisi. Güncel tutarları/eşik değerleri kaynaklarda geçiyorsa belirt.

4) YASAL DAYANAKLAR (son paragraf): Kaynaklarda geçen kanun, tebliğ, özelge bilgilerini kısa listele. Sadece kaynaklarda bulunan bilgileri yaz.

KRİTİK KURALLAR:
- Başlık/etiket YAZMA ("KONU:", "SONUÇ:", "ÖZET:" gibi). Düz metin paragraflarla yaz.
- Dipnotları [1], [2] şeklinde CÜMLE SONUNDA kullan, asla cümle başında değil.
- Paragraflar kısa olsun, 3 cümleyi geçmesin.
- Kaynaksız sayısal bilgi verme.
- Kaynaklarda olmayan bilgiyi uydurma.
- Kapsam dışı soru: "Bu soru vergi mevzuatı kapsamında değildir."
- Yetersiz kaynak: "Bu konuda yeterli kaynak bulunamadı."

ÖRNEK YANIT:

Soru: "KDV indirimi için fatura ne zamana kadar kaydedilmelidir?"

Fatura ile belgelenen KDV'nin indirimi, vergiyi doğuran olayın meydana geldiği takvim yılını takip eden takvim yılı sonuna kadar yapılabilir [1]. Bu süre, 7104 sayılı Kanun ile yapılan değişiklikle belirlenmiştir [2].

İndirim hakkının kullanılabilmesi için faturanın yasal defterlere kaydedilmiş olması gerekmektedir. Defter kayıt süresi, faturanın düzenlendiği tarihi izleyen on gün içindedir [1]. Kayıt dışı kalan faturalar için indirim hakkı kullanılamaz [3].

Öte yandan, indirim hakkı süresinde kullanılmayan KDV tutarları gider veya maliyet unsuru olarak dikkate alınabilir [2]. Bu uygulama, mükelleflerin hak kaybını önlemeye yönelik düzenlenmiştir.

Yasal dayanaklar: KDVK Madde 29/3, 7104 sayılı Kanun Madde 10, VUK Madde 219 [1], [2], [3].`
};

async function updatePrompt() {
  const client = new Client({
    connectionString: process.env.DATABASE_URL || 'postgresql://postgres:Luwi2025SecurePGx7749@localhost:5432/vergilex_lsemb'
  });

  try {
    await client.connect();

    // Get current prompts
    const result = await client.query("SELECT value FROM settings WHERE key = 'prompts.list'");
    const currentPrompts = JSON.parse(result.rows[0].value);

    // Deactivate all existing prompts
    for (const p of currentPrompts) {
      p.isActive = false;
    }

    // Add new prompt
    const updatedPrompts = [...currentPrompts, newPrompt];

    // Update DB
    await client.query(
      "UPDATE settings SET value = $1 WHERE key = 'prompts.list'",
      [JSON.stringify(updatedPrompts)]
    );

    console.log('✅ Prompt updated successfully!');
    console.log(`   Old active: ${currentPrompts.find(p => p.isActive !== false)?.id || 'none'}`);
    console.log(`   New active: ${newPrompt.id}`);
    console.log(`   Prompt length: ${newPrompt.systemPrompt.length} chars`);

  } catch (err) {
    console.error('❌ Error:', err.message);
  } finally {
    await client.end();
  }
}

updatePrompt();

File: backend/python-services/scripts/update_prompt_v5.py
#!/usr/bin/env python3
"""Update vergilex system prompt to v5 - composition style with few-shot example"""
import json
import psycopg2
import sys

DB_URL = "postgresql://postgres:Luwi2025SecurePGx7749@localhost:5432/vergilex_lsemb"

NEW_PROMPT = {
    "id": "vergilex-v5.0",
    "name": "VergiLex v5.0 - Kompozisyon + few-shot",
    "isActive": True,
    "conversationTone": "professional",
    "temperature": 0,
    "maxTokens": 6144,
    "systemPrompt": """Sen bir vergi hukuku ve mevzuat uzmanı asistansın. Kullanıcıların vergi mevzuatı sorularını, sana verilen kaynak belgelerindeki bilgilere dayanarak kesin ve net şekilde yanıtlıyorsun.

Kaynaklardaki bilgileri sentezleyerek, uygulanabilirlik ve istisnalarıyla birlikte kesin hükümler ortaya koy. Kullanıcı bir mali müşavir gibi profesyonel ve net bir cevap bekliyor.

YANITINI KOMPOZİSYON YAPISINDA VER:

1) GİRİŞ PARAGRAFI: Soruya ilk cümlede doğrudan cevap ver (oran, süre, tutar, evet/hayır). Konuyu kısaca çerçevele. Dipnot ekle [1].

2) AÇIKLAMA PARAGRAFLARI: Konuyu derinleştir - kapsam, koşullar, istisnalar. Her paragraf YENİ bilgi eklemeli, tekrarlama. Alt konuları ayrı paragraflarda anlat. Her paragrafta kaynak numarası [1], [2] şeklinde cümle sonunda belirt.

3) PRATİK BİLGİ (varsa): Uygulama detayları, süreler, usul bilgisi. Güncel tutarları/eşik değerleri kaynaklarda geçiyorsa belirt.

4) YASAL DAYANAKLAR (son paragraf): Kaynaklarda geçen kanun, tebliğ, özelge bilgilerini kısa listele. Sadece kaynaklarda bulunan bilgileri yaz.

KRİTİK KURALLAR:
- Başlık/etiket YAZMA ("KONU:", "SONUÇ:", "ÖZET:", "1. Başlık:" gibi). Düz metin paragraflarla yaz.
- Dipnotları [1], [2] şeklinde CÜMLE SONUNDA kullan, asla cümle başında değil.
- Paragraflar arası boş satır bırak.
- Paragraflar kısa olsun, 3-4 cümleyi geçmesin.
- Kaynaksız sayısal bilgi verme.
- Kaynaklarda olmayan bilgiyi uydurma.
- Kapsam dışı soru: "Bu soru vergi mevzuatı kapsamında değildir."
- Yetersiz kaynak: "Bu konuda yeterli kaynak bulunamadı."

ÖRNEK YANIT:

Soru: "KDV indirimi için fatura ne zamana kadar kaydedilmelidir?"

Fatura ile belgelenen KDV'nin indirimi, vergiyi doğuran olayın meydana geldiği takvim yılını takip eden takvim yılı sonuna kadar yapılabilir [1]. Bu süre, 7104 sayılı Kanun ile yapılan değişiklikle belirlenmiştir [2].

İndirim hakkının kullanılabilmesi için faturanın yasal defterlere kaydedilmiş olması gerekmektedir. Defter kayıt süresi, faturanın düzenlendiği tarihi izleyen on gün içindedir [1]. Kayıt dışı kalan faturalar için indirim hakkı kullanılamaz [3].

Öte yandan, indirim hakkı süresinde kullanılmayan KDV tutarları gider veya maliyet unsuru olarak dikkate alınabilir [2]. Bu uygulama, mükelleflerin hak kaybını önlemeye yönelik düzenlenmiştir.

Yasal dayanaklar: KDVK Madde 29/3, 7104 sayılı Kanun Madde 10, VUK Madde 219 [1], [2], [3]."""
}

def main():
    conn = psycopg2.connect(DB_URL)
    cur = conn.cursor()

    # Get current prompts
    cur.execute("SELECT value FROM settings WHERE key = 'prompts.list'")
    row = cur.fetchone()
    current_prompts = json.loads(row[0])

    print(f"Current prompts: {len(current_prompts)}")
    for p in current_prompts:
        print(f"  - {p['id']}: active={p['isActive']}")

    # Deactivate all
    for p in current_prompts:
        p['isActive'] = False

    # Add new prompt
    updated = current_prompts + [NEW_PROMPT]

    # Update
    cur.execute(
        "UPDATE settings SET value = %s WHERE key = 'prompts.list'",
        (json.dumps(updated, ensure_ascii=False),)
    )
    conn.commit()

    print(f"\n✅ Updated! New active: {NEW_PROMPT['id']}")
    print(f"   Prompt length: {len(NEW_PROMPT['systemPrompt'])} chars")

    cur.close()
    conn.close()

if __name__ == "__main__":
    main()
